### PR TITLE
fix build against LLVM master: clEnumValEnd removed

### PR DESCRIPTION
**Pull request by rainers**
Not sure if has landed in circleCI already, but probably will soon.

### CODE REVIEW

File: gen/llvmhelpers.cpp [hunk 1]
```diff
@@ -44,19 +45,24 @@
 
 #include "llvm/Support/CommandLine.h"
 
+#if LDC_LLVM_VER >= 400
+#define _clEnumValN llvm::clEnumValN
+#else
+#define _clEnumValN clEnumValN
+#endif
```
Comment (kinke) [inline gen/llvmhelpers.cpp]: It has landed. :) Is this `_clEnumValN` really required? Would
```C++
#if LDC_LLVM_VER >= 400
using llvm::clEnumValN;
#endif
```
not work?

**Comment by kinke** [comment]
Oh sry seems like I can't simply remove my test commit anymore. :(

**Comment by kinke** [comment]
`llvm-profdata-4.0.cpp` needs to be synced as well.

**Comment by rainers** [comment]
Updated.
I wonder what I have to do (or stop doing) to also build the profile and cache stuff on windows. Does it need a special build of LLVM? I only see the tests continue to fail...

**Comment by kinke** [comment]
> I wonder what I have to do (or stop doing) to also build the profile and cache stuff on windows.

The `all` target includes those tools (at least with ninja).

File: gen/cl_helpers.h [hunk 2]
```diff
@@ -206,4 +206,16 @@ class StringsAdapter {
 };
 }
 
+
+#if LDC_LLVM_VER >= 400
+#define clEnumValues llvm::cl::values
```
Comment (JohanEngelen) [inline gen/cl_helpers.h]: would `using clEnumValues = llvm::cl::values;` work?

File: gen/llvmhelpers.cpp (hunk 1 above)
Comment (JohanEngelen) [inline gen/llvmhelpers.cpp]: Could you rename `_clEnumValN` to ALL_CAPS ?

File: gen/cl_helpers.h (hunk 2 above)
Comment (rainers) [inline gen/cl_helpers.h]: No, the compiler complains that using only works for types.

**Comment by rainers** [comment]
> The all target includes those tools (at least with ninja).

Thanks, it helped me discover the outdated list of targets that my build script used. (With MULTILIB, I have to build the 32-bit libs with different environment settings.)

**Comment by JohanEngelen** [comment]
@rainers I think you should just do whatever you think is best, and then merge this thing asap :-)

File: gen/llvmhelpers.cpp [hunk 3]
```diff
@@ -44,19 +45,23 @@
 
 #include "llvm/Support/CommandLine.h"
 
+#if LDC_LLVM_VER >= 400
+// trick clEnumValN() into prepending the llvm namesspace 
+namespace cl { using OptionEnumValue = llvm::cl::OptionEnumValue; }
```
Comment (rainers) [inline gen/llvmhelpers.cpp]: This workaround can be removed with the next update to LLVM: https://github.com/llvm-mirror/llvm/commit/15a3555c6d39d2b5926b306e4d24bce0cf4d390b

**Pull request merged**